src/module/functions/mod.rs

```diff
@@ -350,7 +350,7 @@ impl Module {
         // necessary and it's a bottleneck!
         let mut bodies = Vec::with_capacity(functions.len());
         for (i, (body, mut validator)) in functions.into_iter().enumerate() {
-            let index = (num_imports + i) as u32;
+            let index = (indices.num_fun_imports + i) as u32;
```

Comment:
I'd be curious to know the reason this is necessary - which is incorrect above - `funcs.arena.len()` or `functions.len()`? I'm also not particularly confident this covers all index changes necessary to support the feature, if you do want to land this it would help to provide a little more analysis / feedback on why you think this is enough to make this PR complete.

Reply:
I did this because `functions.len()` is the length of the current module's functions, which includes the total amount of imports being merged *and* functions previously in the module. Previously, `functions.len()` worked because `functions` started out empty, but with merging, `functions` can contain values before imports.

Reply:
Thanks for clarifying. Are there any other cases where imports being added later might affect indexing, or do you feel pretty confident this is the only one?